graylog2-server/src/main/java/org/graylog2/datanode/DatanodeUpgradeService.java

```diff
@@ -0,0 +1,136 @@
+        final boolean clusterReadyForUpgrade = clusterHealthy && shardReplicationEnabled;
+
+        return new DatanodeUpgradeStatus(serverVersion,
+                clusterState,
+                clusterHealthy,
+                shardReplicationEnabled,
+                enrichData(upToDateDataNodes, clusterState, serverVersion, clusterReadyForUpgrade),
+                enrichData(toUpgradeDataNodes, clusterState, serverVersion, clusterReadyForUpgrade)
+        );
+    }
+
+    private List<DataNodeInformation> enrichData(List<DataNodeDto> nodes, ClusterState clusterState, Version serverVersion, boolean clusterReadyForUpgrade) {
+        final Comparator<DataNodeInformation> comparator = Comparator.comparing(DataNodeInformation::upgradePossible)
+                .reversed()
+                .thenComparing(DataNodeInformation::nodeName);
+
+        AtomicInteger upgradeableCounter = new AtomicInteger(1);
+
+        return nodes.stream()
+                .sorted(Comparator.nullsLast(Comparator.comparing(DataNodeDto::getHostname)))
+                .map(n -> enrichNodeInformation(n, nodes, clusterState, serverVersion, clusterReadyForUpgrade, upgradeableCounter))
+                .sorted(comparator)
+                .collect(Collectors.toList());
+    }
+
+    @Nonnull
+    private static DataNodeInformation enrichNodeInformation(DataNodeDto node, List<DataNodeDto> toUpgradeDataNodes, ClusterState clusterState, Version serverVersion, boolean clusterReadyForUpgrade, AtomicInteger upgradeableCounter) {
+        final Optional<Node> opensearchInformation = clusterState.findByHostname(node.getHostname());
+
+
+        final boolean managerNode = opensearchInformation.map(Node::name)
+                .map(n -> n.equals(clusterState.managerNode().name()))
+                .orElse(false);
+
+        final boolean isLatestVersion = isVersionEqualIgnoreBuildMetadata(node.getDatanodeVersion(), serverVersion);
+
+        boolean upgradeTechnicallyPossible = clusterReadyForUpgrade && !isLatestVersion && (!managerNode || toUpgradeDataNodes.size() == 1);
+        // we want to mark only one node as ready for upgrade, guiding the user one by one. The upgradeableCounter keeps the overview
+        boolean upgradeEnabled = upgradeTechnicallyPossible && upgradeableCounter.getAndDecrement() == 1;
+
+        return new DataNodeInformation(
+                opensearchInformation.map(Node::name).orElse(node.getHostname()),
+                node.getDataNodeStatus(),
+                node.getDatanodeVersion(),
+                node.getHostname(),
+                opensearchInformation.map(Node::ip).orElse(null),
+                opensearchInformation.map(Node::version).orElse(null),
+                opensearchInformation.map(Node::roles).orElse(null),
+                upgradeEnabled,
+                managerNode);
+    }
+
+    protected static boolean isVersionEqualIgnoreBuildMetadata(String datanodeVersion, Version serverVersion) {
```

Comment:
This can be replaced with DataNodeDto.isCompatibleWithVersion I think

Reply:
I rather copy-pasted the method as it makes it more explicit which versions are we comparing, how and where is the information coming from. Especially because we are then providing the server version in the API response.

If we want to unify the method, then I'd vote for changing the `DataNodeDto.isCompatibleWithVersion`, which is hiding some complexity.